### PR TITLE
buttons for json editor

### PR DESCRIPTION
**Pull request by larkob**
This PR introduces buttons for all commands in the right panel of the json editor.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3220/pr-test (or any other room on that server)

**Review by 96LawDawg — APPROVED** [review]
Well, this is just an awesome improvement.  Well done, Lars.  Tested and it works.  Of course, I could not possibly test every button combination (I could but I don't have time for that).  But that random sample I did test worked perfectly.

**Comment by robartsd** [comment]
This will be really helpful when there is a conflict between JSON editor shortcuts and other shortcuts. Chrome on my Debian desktop opens a tab in a new window with ctrl-shift-T; without this, the only way for me to get back to tree view is deleting a widget (ctrl-shift-A to add a widget, then ctrl-shift-R to remove it). I also had a regEx search extension mapped to ctrl-shift-F, but I remapped it to alt-shift-F.

**Comment by larkob** [comment]
I've made a significant update:
* buttons have icon (where appropriate)
* common buttons (those without specific context) are now displayed in top row
* clicking on any widget in tree view directly opens the widget

I also found a bug related to `jeGetLastKey`. When the context is in an empty property like `faces: {}` the editor crashed. I added a `toString()` before the `match` to fix the bug.

File: client/js/jsonedit.js [hunk 1]
```diff
@@ -818,7 +864,12 @@ function jeShowCommands() {
           if(!command.currentKey && !usedKeys[key])
             command.currentKey = key;
         usedKeys[command.currentKey] = true;
-        commandText += `Ctrl-${command.currentKey}: ${name.replace(command.currentKey, '<b>' + command.currentKey + '</b>')}\n`;
+        var displayKey = command.currentKey;
+        if (displayKey == 'ArrowUp')
+          displayKey = '⬆';
+        if (displayKey == 'ArrowDown')
+          displayKey = '⬇';
```
Comment (ArnoldSmith86) [inline client/js/jsonedit.js:871]: This is fine for now. If there are more, this should be more like `let displayKey = { ArrowUp: '⬆', ArrowDown: '⬇'}[command.currentKey] || command.currentKey;` (not necessarily on one line).

File: client/js/jsonedit.js [hunk 2]
```diff
@@ -830,6 +881,23 @@ function jeShowCommands() {
   if(jeSecondaryWidget)
     commandText += `\n\n${jeSecondaryWidget}\n`;
   $('#jeCommands').innerHTML = commandText;
+  var buttons = [... document.querySelectorAll('#jeCommands>button')];
+  if (buttons.length > 0) {
+    buttons.forEach((e) => {
+      if (e != undefined) document.getElementById(e.id).addEventListener('click', clickButton);
+    });
+  }
+}
```
Comment (ArnoldSmith86) [inline client/js/jsonedit.js:890]: This looks way too complicated to me. No need to check for `length > 0` (`forEach` should be fine). No idea why you would expect `undefined` values inside the callback. And `e` already is a DOM element. So this should be the same:

```Javascript
let buttons = [... document.querySelectorAll('#jeCommands>button')];
buttons.forEach(e => e.addEventListener('click', clickButton));
```

There is actually a helper function `$a` for `document.querySelectorAll` though:

```Javascript
for(const e of $a('#jeCommands>button'))
  e.addEventListener('click', clickButton);
```

But there's even a helper function for that:

```Javascript
on('#jeCommands>button', 'click', clickButton);
```

_(Didn't test any of these.)_

File: client/js/jsonedit.js [hunk 3]
```diff
@@ -830,6 +881,23 @@ function jeShowCommands() {
...
+
+const clickButton = function (event) {
+  jeCommands.find(o => o.id == event.currentTarget.id).call();
+  if (jeContext != 'macro') {
+    jeGetContext();
+    if(jeWidget && !jeJSONerror)
+      jeApplyChanges();
+    if (jeContext[0] == '###SELECT ME###')
+      jeGetContext();
```
Comment (ArnoldSmith86) [inline client/js/jsonedit.js:899]: This doesn't look like it belongs here but I'm pretty sure it's fixing a bug that I also noticed before so whatever. :D

If I didn't miss anything this is only called for clicks and not hotkeys, right? Feels like something that should be the same either way.

File: client/js/jsonedit.js [hunk 4]
```diff
@@ -858,8 +926,12 @@ window.addEventListener('mousemove', function(e) {
 window.addEventListener('mouseup', function(e) {
   if(!jeEnabled)
     return;
-  if(e.target == $('#jeText'))
+  if(e.target == $('#jeText') && jeContext != 'macro') {
     jeGetContext();
+    if (jeContext[0] == 'Tree' && jeContext[1] != undefined)
+      jeCommands.find(o => o.id == 'je_openWidgetById').call();
+      jeGetContext();
+  }
```
Comment (ArnoldSmith86) [inline client/js/jsonedit.js:934]: No idea what this change does. But you probably have your reasons. Oh wait, it opens a widget when you click on the tree?

Looks like either the indentation is wrong or it's missing curly brackets though..

**Review by ArnoldSmith86 — APPROVED** [review]
I don't see any real issues.

You could have also left your icons in there and just skipped anything but numbers and letters when going through hotkey suggestions.

File: client/js/jsonedit.js [hunk 5]
```diff
@@ -22,13 +22,15 @@ const jeOrder = [ 'type', 'id#', 'parent', 'deck', 'cardType', 'owner#', 'x*', '
 
 const jeCommands = [
   {
+    id: 'je_toggleBoolean',
     name: 'toggle boolean',
-    context: '.*"(true|false)"',
+    context: '.*(true|false)',
```
[the remark this answers is not in the document]
Comment (larkob) [reply client/js/jsonedit.js]: I couldn't find a difference in function, so I reverted it back to how you did this.

File: client/js/jsonedit.js (hunk 1 above)
Comment (larkob) [reply client/js/jsonedit.js]: Yes, good point. I actually added a third arrow key for the "convert to applyVariables" and used your suggestion

File: client/js/jsonedit.js (hunk 2 above)
Comment (larkob) [reply client/js/jsonedit.js]: Yes, the `on` is much easier and worked perfectly.

File: client/js/jsonedit.js (hunk 4 above)
Comment (larkob) [reply client/js/jsonedit.js]: Right, I've added the brackets. It's odd that jeGetContext() needs to be called twice in the tree view, but I couldn't find any other way to make this work.

**Comment by ArnoldSmith86** [comment]
Looks like this for me (Firefox 84 on Linux):

Screenshot from 2021-02-28 14-58-55

**Comment by ArnoldSmith86** [comment]
Looks like this if I add `border: 0`:

Screenshot from 2021-02-28 15-55-11

File: client/js/jsonedit.js (hunk 3 above)
Comment (larkob) [reply client/js/jsonedit.js:881]: yes, clicks only. that bug didn't occur using keys

**Pull request merged**